spot-client/src/common/remote-control/P2PSignalingClient.js

```diff
@@ -0,0 +1,156 @@
+
+        /**
+         * @typedef {Object} PromiseLike
+         * @property {Function} resolve - hooked up to Promise's resolve method.
+         * @property {Function} reject - hooked up to Promise's reject method.
+         */
+        /**
+         * Map stores command requests sent over P2P signaling in order to resolve original promises when ack is
+         * received.
+         * @type {Map<number, PromiseLike>}
+         * @private
+         */
+        this._p2pSignalingRequests = new Map();
+    }
+
+    /**
+     * A convenience selector to find the first {@link PeerConnection}.
+     *
+     * @returns {?PeerConnection}
+     */
+    _getFirstPeerConnection() {
+        return this._peerConnections.values().next().value;
+    }
+
+    /**
+     * Checks if the connection is ready to send commands.
+     *
+     * @returns {boolean}
+     */
+    isReady() {
+        const first = this._getFirstPeerConnection();
+
+        return first && first.isDataChannelActive();
+    }
+
+    /**
+     * Client processing for data channel messages. A client processes only acks.
+     *
+     * @param {string} remoteAddress - The remote address that sent the message.
+     * @param {Object} msg - JSON object received in message content.
+     * @private
+     * @returns {void}
+     */
+    _processDataChannelMessage(remoteAddress, msg) {
+        const { command, requestId } = msg;
+
+        if (command === 'ack' && requestId) {
+            const commandPromise = this._p2pSignalingRequests.get(requestId);
+
+            if (commandPromise) {
+                commandPromise.resolve();
+            } else {
+                logger.warn('No P2P command promise exists', { requestId });
+            }
+
+            this._p2pSignalingRequests.delete(requestId);
+        } else if (msg.command === 'status' && msg.newStatus) {
+            this.emit(P2PSignalingClient.SPOT_TV_STATUS_UPDATE, remoteAddress, msg.newStatus);
+        } else {
+            super._processDataChannelMessage(remoteAddress, msg);
+        }
+    }
+
+    /**
+     * Callback triggered by {@link PeerConnection} whenever data channel's readiness state changes.
+     *
+     * @param {string} remoteAddress - The remote address associated with source {@link PeerConnection}.
+     * @param {boolean} isReady - Is it ready or not.
+     * @protected
+     * @returns {void}
+     */
+    _onDCReadyStateChanged(remoteAddress, isReady) {
+        super._onDCReadyStateChanged(remoteAddress, isReady);
+
+        if (!this.isReady()) {
+
+            // NOTE: It might be possible to give it some time to reconnect
+            const count = this._p2pSignalingRequests.size;
+
+            count && logger.log('Rejecting signaling promises', { count });
+
+            for (const promise of this._p2pSignalingRequests.values()) {
+                promise.reject('P2P signaling channel closed');
+            }
+
+            this._p2pSignalingRequests.clear();
+        }
+    }
```

Comment:
Is there a mechanism built in to try to recreate the data-channel or peer connection if either become closed?

Reply:
Currently only if either TV or remote reconnects.

Reply:
But I think this is good enough for the start. The reconnect logic is on the higher level which means whenever network connectivity is lost p2p will be restarted as well. I'd say we add analytics to see when remote is connected over p2p and the figure it out from there if it turns out there are corner cases worth chasing.